File: neural/cpp/Neural/Neural/src/activations.cc
#include "activations.h"

_differentiable Neural::Activations::Linear(Eigen::MatrixXd &z)
{
    _differentiable a;

    a.function = z;
    a.derivative = Eigen::MatrixXd::Ones(z.rows(), z.cols());

    return a;
}

_differentiable Neural::Activations::Sigmoid(Eigen::MatrixXd &z)
{
    _differentiable a;

    a.function = 1 / (1 + Eigen::exp(z.array()));
    a.derivative = a.function.array() * (1 - a.function.array());

    return a;
}

_differentiable Neural::Activations::ReLU(Eigen::MatrixXd &z)
{
    _differentiable a;

    Eigen::MatrixXd F(z.rows(), z.cols());
    Eigen::MatrixXd dF(z.rows(), z.cols());

    for (int i = 0; i < z.rows(); i++)
    {
        for (int j = 0; j < z.cols(); j++)
        {
            if (z(i, j) <= 0)
            {
                F(i, j) = 0;
                dF(i, j) = 0;
            }
            else
            {
                F(i, j) = z(i, j);
                dF(i, j) = 1;
            }
        }
    }

    a.function = F;
    a.derivative = dF;

    return a;
}

_differentiable Neural::Activations::Softmax(Eigen::MatrixXd &z)
{
    _differentiable a;

    a.function = Eigen::exp(z.array() - z.maxCoeff());
    a.function /= z.sum();
    a.derivative = a.function.array() * (1 - a.function.array());

    return a;
}

_differentiable Neural::Activations::Tanh(Eigen::MatrixXd &z)
{
    _differentiable a;

    a.function = z.array().tanh();
    a.derivative = Eigen::pow(1 / Eigen::cosh(z.array()), 2);

    return a;
}

File: neural/python/activations.py
import numpy as np

def linear(z, derivative=False):
    if derivative:
        return 1
    return z

def sigmoid(z, derivative=False):
    f = 1 / (1 + np.exp(-z))

    if derivative:
        return f * (1 - f)
    
    return f

def relu(z, derivative=False, epsilon=1e-16):
    f = np.maximum(z + epsilon, 0)

    if derivative:
        if np.sum(f) != 0:
            return 0
        else:
            return 1
    
    return f

def softmax(z, derivative=False):
    f = np.exp(z - np.max(z, axis=1, keepdims=True))
    f /= np.sum(f, axis=1, keepdims=True)

    if derivative:
        return f * (1 - f)
    
    return f

def tanh(z, derivative=False):
    if derivative:
        return (1 / np.cosh(z)) ** 2
    
    return np.tanh(z)